File: src/main/java/io/dfarquharson/euler/java/problems/Problem2.java
package io.dfarquharson.euler.java.problems;

import io.dfarquharson.euler.java.utils.FibonacciSequence;

import java.util.function.Supplier;

public class Problem2 implements Supplier<String> {

    @Override
    public String get() {
        return "Problem 2 result: %d".formatted(
                new FibonacciSequence()
                        .get()
                        .filter(x -> x % 2 == 0)
                        .takeWhile(x -> x < 4_000_000)
                        .sum()
        );
    }

    public static void main(String[] args) {
        System.out.println(new Problem2().get());
    }

}

File: src/main/kotlin/io/dfarquharson/euler/kotlin/Problem2.kt
package io.dfarquharson.euler.kotlin

fun main() {
    println(Problem2.run())
}

object Problem2 {

    fun sumOfEvenFibsAtOrBelowN(n: Int): Int {
        return Fibonacci.fibs()
            .takeWhile { it <= n }
            .filter { isEven(it) }
            .sum()
    }

    fun isEven(n: Int): Boolean {
        return n % 2 == 0
    }

    fun run(): Int {
        return sumOfEvenFibsAtOrBelowN(4_000_000)
    }

}

object Fibonacci {

    fun fibs(): Sequence<Int> = sequence {
        var elements: Pair<Int, Int> = Pair(1, 2)
        while (true) {
            yield(elements.first)
            elements = Pair(elements.second, elements.first + elements.second)
        }
    }

}
